repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py

```
@@ -0,0 +1,669 @@
+            pool.sort()
+            if [number for number, source in pool] != [0, 1, 2, 3]:
+                # only exact group of 4 servers can be converted, nothing to do here
+                continue
+            # verify that parameters are the same for all servers in the pool
+            if not all([p[1]["options"] == pool[0][1]["options"] for p in pool]):
+                break
+            remove_servers.update([pool[i][1]["address"] for i in [0, 1, 3]])
+            pool[2][1]["type"] = "pool"
+
+        processed_sources = []
+        for source in time_sources:
+            if source["type"] == "server" and source["address"] in remove_servers:
+                continue
+            processed_sources.append(source)
+        return processed_sources
+
+    def get_chrony_conf_sources(self):
+        conf = ""
+
+        if self.step_tickers:
+            conf += "# Specify NTP servers used for initial correction.\n"
+            conf += "initstepslew 0.1 {}\n".format(" ".join(self.step_tickers))
+            conf += "\n"
+
+        conf += "# Specify time sources.\n"
+
+        for source in self.get_processed_time_sources():
+            address = source["address"]
+            if address.startswith("127.127."):
+                if address.startswith("127.127.1."):
+                    continue
+                assert False
+            else:
+                conf += "{} {}".format(source["type"], address)
+                for option in source["options"]:
+                    if option[0] in ["minpoll", "maxpoll", "version", "key",
+                                     "iburst", "noselect", "prefer", "xleave"]:
+                        conf += " {}".format(" ".join(option))
+                    elif option[0] == "burst":
+                        conf += " presend 6"
+                    elif option[0] == "true":
+                        conf += " trust"
+                    else:
+                        assert False
```

Ditto here, please raise an exception with a meaningful message.

---

I don't think a different exception would be appropriate here. I have added a comment to make it more clear what the assertion is supposed to catch.

---

We try to provide all the necessary context to the users to fix the actual problem, why am I supposed to read code when the actual problem is in my data?

---

Users will not see this. Please have a look at the actual code.